indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java

```diff
@@ -0,0 +1,104 @@
+ */
+
+package io.druid.indexing.overlord.setup;
+
+import com.google.common.collect.ImmutableMap;
+import io.druid.indexing.common.task.Task;
+import io.druid.indexing.overlord.ImmutableWorkerInfo;
+import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class WorkerSelectUtils
+{
+  private WorkerSelectUtils()
+  {
+    // No instantiation.
+  }
+
+  /**
+   * Helper for {@link WorkerSelectStrategy} implementations.
+   *
+   * @param allWorkers     map of all workers, in the style provided to {@link WorkerSelectStrategy}
+   * @param affinityConfig affinity config, or null
+   * @param workerSelector function that receives a list of eligible workers: version is high enough, worker can run
+   *                       the task, and worker satisfies the affinity config. may return null.
+   *
+   * @return selected worker from "allWorkers", or null.
+   */
+  @Nullable
+  public static ImmutableWorkerInfo selectWorker(
+      final Task task,
+      final Map<String, ImmutableWorkerInfo> allWorkers,
+      final WorkerTaskRunnerConfig workerTaskRunnerConfig,
+      @Nullable final AffinityConfig affinityConfig,
+      final Function<ImmutableMap<String, ImmutableWorkerInfo>, ImmutableWorkerInfo> workerSelector
+  )
+  {
+    // Workers that could potentially run this task, ignoring affinityConfig.
+    final Map<String, ImmutableWorkerInfo> runnableWorkers = allWorkers
+        .values()
+        .stream()
+        .filter(worker -> worker.canRunTask(task)
+                          && worker.isValidVersion(workerTaskRunnerConfig.getMinWorkerVersion()))
+        .collect(Collectors.toMap(w -> w.getWorker().getHost(), Function.identity()));
+
+    if (affinityConfig == null) {
+      return workerSelector.apply(ImmutableMap.copyOf(runnableWorkers));
+    } else {
+      // Workers not assigned to any affinity pool at all.
+      final Map<String, ImmutableWorkerInfo> nonAffinityWorkers = runnableWorkers
+          .entrySet()
+          .stream()
+          .filter(entry -> !affinityConfig.getAffinityWorkers().contains(entry.getKey()))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+      // Workers assigned to the affinity pool for our task.
+      final Set<String> dataSourceWorkers = affinityConfig.getAffinity().get(task.getDataSource());
+
+      if (dataSourceWorkers == null) {
+        // No affinity config for this dataSource; use non-affinity workers.
+        return workerSelector.apply(ImmutableMap.copyOf(nonAffinityWorkers));
+      } else {
+        final Map<String, ImmutableWorkerInfo> dataSourceWorkerMap = runnableWorkers
+            .entrySet()
+            .stream()
+            .filter(entry -> dataSourceWorkers.contains(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
```

Comment:
Could be as concise as, i.e. `Map<String, ImmutableWorkerInfo> filtered = Maps.filterKeys(runnableWorkers, s -> dataSourceWorkers.contains(s));`

See `com.google.common.collect.Maps`

Reply:
Changed it.